### PR TITLE
feat(providers): add FallbackBoundModel — BoundModel-level provider failover chain

### PR DESCRIPTION
**Pull request by xfgong**
## Summary

- Add `FallbackBoundModel` to `cubepi/providers/fallback.py` — a standalone frozen dataclass wrapping an ordered `chain: tuple[BoundModel, ...]`. On a retriable error (default: `RateLimited`, `ProviderUnavailable`, `ContextLengthExceeded`) or a first-event stream error, it transparently tries the next model in the chain.
- Export `FallbackBoundModel` and `DEFAULT_TRIGGER_ERRORS` from `cubepi.providers` and `cubepi` top-level.
- Widen `Agent.__init__` and agent loop function signatures to accept `BoundModel | FallbackBoundModel`.

## Design

`FallbackBoundModel` sits at the `BoundModel` level (not the `Provider` level) because the fallback chain contains heterogeneous `(provider, spec)` pairs — there is no single model spec to pass from outside. `provider`/`spec` properties proxy `chain[0]` so tracing/billing code that reads `agent._model.provider`/`agent._model.spec` works unchanged.

Failover strategy: first-event-only (approach A). `stream()` peeks at the first `StreamEvent`; if it is `type="error"`, or if `await bound.stream()` raises a typed error in `trigger_errors`, the next model is tried. Mid-stream errors are forwarded as-is.

`on_failover(failed, next_model, error)` optional callback for billing/metrics hooks (cubebox will wire `billing.record_fallback_failure()` here). Callback exceptions are swallowed.

Spec: `dev/specs/2026-06-08-fallback-bound-model.md`
Plan: `dev/plans/2026-06-08-fallback-bound-model.md`

## Breaking

None. Existing `Agent(model=BoundModel(...))` call sites are unaffected — `BoundModel` remains valid. `FallbackBoundModel` is additive.

## Test plan

- [x] `uv run pytest tests/` — 1544 passed, 45 skipped
- [x] `uv run mypy cubepi/` — Success, no issues in 89 source files
- [x] `uv run ruff check cubepi/ tests/` — All checks passed
- [x] `uv run ruff format --check cubepi/ tests/` — 90 files already formatted
- [x] 9 dedicated tests in `tests/providers/test_fallback.py` covering: primary success, trigger-error failover, non-trigger re-raise, first-event-error failover, all-exhausted, swallowed callback exception, `generate()` path, custom `trigger_errors`

### CODE REVIEW

File: cubepi/providers/fallback.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                first = await iterator.__anext__()
+            except StopAsyncIteration:
+                last_error = "stream ended before producing any events"
+                await self._notify(bound, next_bound, last_error, attempt)
+                continue
+
+            if first.type == "error":
+                last_error = first.error_message or "stream error"
+                await self._notify(bound, next_bound, last_error, attempt)
+                continue
```
[the remark this answers is not in the document]
Comment (xfgong) [reply cubepi/providers/fallback.py:152]: Acknowledged — this is a real tension. The spec explicitly chose "always fallback on first-event stream error" because `StreamEvent` carries only a string `error_message` with no structured type, so filtering by `trigger_errors` at the event level is not reliably possible without string-parsing heuristics.

The practical consequence you describe is correct: a `ProviderAuthFailed` that surfaces as a stream event (the normal path for Anthropic/OpenAI) will trigger failover even though `ProviderAuthFailed` is excluded from `DEFAULT_TRIGGER_ERRORS`. This is the documented tradeoff — the alternative (not failing over on first-event errors) would mean most real provider failures never trigger fallback at all, since built-in providers emit errors as events rather than raising from `stream()`.

A `fallback_on_stream_error: bool = True` parameter could give callers strict control at the cost of a slightly broader API surface. Leaving that as a follow-up for now — the current default (always fallback on first-event errors) is the right behavior for the primary use-case of transient outages and rate limits. Skipping this suggestion.

**Comment by xfgong** [comment]
@codex please take another pass — pushed 8c7ebd3 (docs: CHANGELOG, README, recipe, providers guide).

**Comment by xfgong** [comment]
@codex please take another pass — pushed b0c49b5 (test: coverage for empty stream, mid-stream error, generate exhausted, provider/spec properties — 99% coverage on fallback.py).
